metricbeat/module/vsphere/host/host.go

```diff
@@ -92,83 +111,187 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+			if err := pc.Retrieve(ctx, virtualMachineRefs, []string{"name"}, &vms); err != nil {
+				m.Logger().Errorf("Failed to retrieve virtual machine from host: %v", err)
+				continue
 			}
 		}
-		reporter.Event(mb.Event{
-			MetricSetFields: event,
-		})
-	}
 
-	return nil
-}
+		outputVmNames := make([]string, 0, len(vms))
+		for _, vm := range vms {
+			name := strings.ReplaceAll(vm.Name, ".", "_")
+			outputVmNames = append(outputVmNames, name)
+		}
 
-func getNetworkNames(ctx context.Context, c *vim25.Client, ref types.ManagedObjectReference) ([]string, error) {
-	var outputNetworkNames []string
+		networkRefs := make([]types.ManagedObjectReference, 0, len(hs.Network))
+		for _, obj := range hs.Network {
+			if obj.Type == "Network" {
+				networkRefs = append(networkRefs, obj)
+			}
+		}
 
-	pc := property.DefaultCollector(c)
+		var nets []mo.Network
+		if len(networkRefs) > 0 {
+			if err := pc.Retrieve(ctx, networkRefs, []string{"name"}, &nets); err != nil {
+				m.Logger().Errorf("Failed to retrieve network from host: %v", err)
+				continue
+			}
+		}
 
-	var hs mo.HostSystem
-	err := pc.RetrieveOne(ctx, ref, []string{"network"}, &hs)
-	if err != nil {
-		return nil, fmt.Errorf("error retrieving host information: %v", err)
-	}
+		outputNetworkNames := make([]string, 0, len(nets))
+		for _, net := range nets {
+			name := strings.ReplaceAll(net.Name, ".", "_")
+			outputNetworkNames = append(outputNetworkNames, name)
+		}
 
-	if len(hs.Network) == 0 {
-		return nil, errors.New("no networks found")
-	}
+		spec := types.PerfQuerySpec{
+			Entity:     hs.Reference(),
+			MetricId:   metricIDs,
+			MaxSample:  1,
+			IntervalId: 20, // right now we are only grabbing real time metrics from the performance manager
+		}
 
-	var networkRefs []types.ManagedObjectReference
-	for _, obj := range hs.Network {
-		if obj.Type == "Network" {
-			networkRefs = append(networkRefs, obj)
+		// Query performance data
+		samples, err := perfManager.Query(ctx, []types.PerfQuerySpec{spec})
+		if err != nil {
+			m.Logger().Errorf("Failed to query performance data: %v", err)
+			continue
 		}
-	}
 
-	if len(networkRefs) == 0 {
-		return nil, errors.New("no networks found")
-	}
+		if len(samples) == 0 {
+			m.Logger().Debug("No samples returned from performance manager")
+			continue
+		}
 
-	var nets []mo.Network
-	err = pc.Retrieve(ctx, networkRefs, []string{"name"}, &nets)
-	if err != nil {
-		return nil, fmt.Errorf("error retrieving network from host: %v", err)
-	}
+		results, err := perfManager.ToMetricSeries(ctx, samples)
+		if err != nil {
+			m.Logger().Errorf("Failed to convert performance data: %v", err)
+		}
+
+		for _, result := range results[0].Value {
+			if len(result.Value) > 0 {
+				if assignValue, exists := metricMap[result.Name]; exists {
+					*assignValue = result.Value[0] // Assign the metric value to the variable
+				}
+			} else {
+				m.Logger().Debug("Metric ", result.Name, ": No result found")
+			}
+		}
```

Comment:
As suggested by @shmsr, you could further avoid nesting
```suggestion
		for _, result := range results[0].Value {
	               if len(result.Value) == 0 {
		           m.Logger().Debug("Metric ", result.Name, ": No result found")
		           continue
	               } 
	
	               if assignValue, exists := metricMap[result.Name]; exists {
		         *assignValue = result.Value[0] // Assign the metric value to the variable
	              }
                }
```